controllers/playlist.controller.js

```diff
@@ -0,0 +1,170 @@
+const User = require("../models/user-model");
+const Playlist = require("../models/playlist-model");
+const { concat } = require("lodash");
+
+const getPlaylists = async (req, res) => {
+  const playlist = await Playlist.find({});
+  res.status(200).json({ success: true, playlist });
+};
+
+const findUserPlaylist = async (req, res, next, userId) => {
+  try {
+    let user = await User.findOne({ _id: userId });
+    if (!user) {
+      res.status(404).json({
+        success: false,
+        message: "Invalid user! Kindly register to continue",
+      });
+      throw Error("Invalid User");
+    }
+    let playlist = await Playlist.findOne({ userId });
+
+    if (!playlist) {
+      playlist = new Playlist({
+        userId,
+        playlists: [{ name: "Custom 1", videos: [], active: true }],
+      });
+      playlist = await playlist.save();
+    }
+    req.playlist = playlist;
+    next();
+  } catch (error) {
+    console.log(error)
+    return res.status(500).json({
+      success: false,
+      message: "Unable to retrive user's playlists",
+      errorMessage: error.message,
+    });
+  }
+};
+
+const getActivePlaylistItems = async (playlist) => {
+  playlist.playlists = playlist.playlists = playlist.playlists.filter(
+    (list) => list.active
+  );
+  for(let list of playlist.playlists){
+    if(list.videos.length>0){
+      list.videos = list.videos.filter(video => video.active);
+    }
+  }
+  return playlist.playlists;
+};
+
+const getUserPlaylist = async (req, res) => {
+  try {
+    let { playlist } = req;
+    const playlistItems = await getActivePlaylistItems(playlist);
+    return res.status(200).json({ success: true, playlist: playlistItems });
+  } catch (err) {
+    console.log(err)
+    return res.status(500).json({
+      success: false,
+      message: "Unable to retrive the playlist",
+      errMessage: err.message,
+    });
+  }
+};
+
+const createPlaylist = async (req, res) => {
+  const { name, _id } = req.body;
+  let { playlist } = req;
+  let newList = {
+    name,
+    videos: [{ _id, active: true }],
+    active: true,
+  };
+  playlist.playlists = concat(playlist.playlists,newList );
+  let updatedPlaylist = await playlist.save();
+  newList = updatedPlaylist.playlists[updatedPlaylist.playlists.length-1];
+  return res.status(201).json({ success: true, playlist: newList });
+};
+
+const updatePlaylistName = async (req, res) => {
+  const { _id, name } = req.body;
+  let { playlist } = req;
+  for (let list of playlist.playlists) {
+    if (list._id == _id) {
+      list.name = name;
+      break;
+    }
+  }
+  let updatedPlaylist = await playlist.save();
+  updatedPlaylist = await getActivePlaylistItems(updatedPlaylist);
+  return res.status(200).json({ success: true, playlist: updatedPlaylist });
+};
+
+const getVideosInPlaylist =  (playlist, listId) => {
+    let playlistItem = playlist.playlists.find((item) => item._id == listId && item.active);
```

Comment:
the mongoose function `aggregate()` can be used here

Reply:
Will look into it